masonry/src/passes/event.rs

```diff
@@ -0,0 +1,277 @@
+    root_state: &mut WidgetState,
+    event: &PointerEvent,
+) -> Handled {
+    let _span = info_span!("pointer_event").entered();
+    if !event.is_high_density() {
+        debug!("Running ON_POINTER_EVENT pass with {}", event.short_name());
+    }
+
+    root.last_mouse_pos = event.position();
+
+    let target_widget_id = get_target_widget(root, event.position());
+
+    let handled = run_event_pass(
+        root,
+        root_state,
+        target_widget_id,
+        event,
+        |widget, ctx, event| {
+            widget.on_pointer_event(ctx, event);
+        },
+    );
+
+    if !event.is_high_density() {
+        debug!(
+            focused_widget = root.state.focused_widget.map(|id| id.0),
+            handled = handled.is_handled(),
+            "ON_POINTER_EVENT finished",
+        );
+    }
+
+    handled
+}
+
+pub fn root_on_text_event(
+    root: &mut RenderRoot,
+    root_state: &mut WidgetState,
+    event: &TextEvent,
+) -> Handled {
+    let _span = info_span!("text_event").entered();
+    if !event.is_high_density() {
+        debug!("Running ON_TEXT_EVENT pass with {}", event.short_name());
+    }
+
+    let target = root.state.focused_widget;
+
+    let mut handled = run_event_pass(root, root_state, target, event, |widget, ctx, event| {
+        widget.on_text_event(ctx, event);
+    });
+
+    // Handle Tab focus
+    if let TextEvent::KeyboardKey(key, mods) = event {
+        if handled == Handled::No && key.physical_key == PhysicalKey::Code(KeyCode::Tab) {
+            if !mods.shift_key() {
+                root.state.next_focused_widget = root.widget_from_focus_chain(true);
+            } else {
+                root.state.next_focused_widget = root.widget_from_focus_chain(false);
+            }
+            handled = Handled::Yes;
+        }
+    }
+
+    if !event.is_high_density() {
+        debug!(
+            focused_widget = root.state.focused_widget.map(|id| id.0),
+            handled = handled.is_handled(),
+            "ON_TEXT_EVENT finished",
+        );
+    }
+
+    handled
+}
+
+pub fn root_on_access_event(
+    root: &mut RenderRoot,
+    root_state: &mut WidgetState,
+    event: &AccessEvent,
+) -> Handled {
+    let _span = info_span!("access_event").entered();
+    debug!("Running ON_ACCESS_EVENT pass with {}", event.short_name());
+
+    let target = Some(event.target);
+
+    let handled = run_event_pass(root, root_state, target, event, |widget, ctx, event| {
+        widget.on_access_event(ctx, event);
+    });
+
+    debug!(
+        focused_widget = root.state.focused_widget.map(|id| id.0),
+        handled = handled.is_handled(),
+        "ON_ACCESS_EVENT finished",
+    );
+
+    handled
+}
+
+// ---
+
+fn run_event_pass<E>(
+    root: &mut RenderRoot,
+    root_state: &mut WidgetState,
+    target: Option<WidgetId>,
+    event: &E,
+    pass_fn: impl FnMut(&mut dyn Widget, &mut EventCtx, &E),
+) -> Handled {
+    let mut pass_fn = pass_fn;
+
+    let mut target_widget_id = target;
```

Comment:
Nit: you could just use `target`/`pass_fn` directly, by declaring it `mut` in the function signature (or is this intentional?)

Reply:
Done for pass_fn. Renamed target_widget_id for something clearer instead.